In [1]:
import numpy as np
import pandas as pd

import sqlite3

**Database: Chinook**

In [17]:
# Kết nối đến CSDL và tạo cursor
conn = sqlite3.connect('database/Chinook_Sqlite.sqlite')
cursor = conn.cursor()

# 1. Gom nhóm

## 1.1. Mệnh đề `GROUP BY`
(Lưu ý: `GROUP BY` cần kết hợp với `MAX`, `MIN`, `SUM`, `COUNT`, `AVG`)

In [5]:
# Gom nhóm bảng track theo AlbumID 
query = """
        SELECT *
        FROM track
        GROUP BY albumid
        --LIMIT (5)
"""
cursor.execute(query)
fetch_data = cursor.fetchall()
pd.DataFrame(fetch_data, columns=[item[0] for item in cursor.description])

TrackId                                               Name  AlbumId  \
0          1            For Those About To Rock (We Salute You)        1   
1          2                                  Balls to the Wall        2   
2          3                                    Fast As a Shark        3   
3         15                                            Go Down        4   
4         23                                      Walk On Water        5   
..       ...                                                ...      ...   
342     3499  Pini Di Roma (Pinien Von Rom) \ I Pini Della V...      343   
343     3500  String Quartet No. 12 in C Minor, D. 703 "Quar...      344   
344     3501               L'orfeo, Act 3, Sinfonia (Orchestra)      345   
345     3502  Quintet for Horn, Violin, 2 Violas, and Cello ...      346   
346     3503                                      Koyaanisqatsi      347   

     MediaTypeId  GenreId                                           Composer  \
0              1        1          Angus Young, Malcolm Young, Brian Johnson   
1              2        1                                               None   
2              2        1  F. Baltes, S. Kaufman, U. Dirkscneider & W. Ho...   
3              1        1                                              AC/DC   
4              1        1   Steven Tyler, Joe Perry, Jack Blades, Tommy Shaw   
..           ...      ...                                                ...   
342            2       24                                               None   
343            2       24                                     Franz Schubert   
344            2       24                                 Claudio Monteverdi   
345            2       24                            Wolfgang Amadeus Mozart   
346            2       10                                       Philip Glass   

     Milliseconds     Bytes  UnitPrice  
0          343719  11170334       0.99  
1          342562   5510424       0.99  
2          230619   3990994       0.99  
3          331180  10847611       0.99  
4          295680   9719579       0.99  
..            ...       ...        ...  
342        286741   4718950       0.99  
343        139200   2283131       0.99  
344         66639   1189062       0.99  
345        221331   3665114       0.99  
346        206005   3305164       0.99  

[347 rows x 9 columns]

In [6]:
# Kiểm tra dữ liệu bảng "track"
query = """
        SELECT COUNT(trackID) as Count_Track, GenreID
        FROM track
        GROUP BY GenreID
        ORDER BY Count_Track DESC
        --LIMIT (5)
"""
cursor.execute(query)
fetch_data = cursor.fetchall()
pd.DataFrame(fetch_data, columns=[item[0] for item in cursor.description])

Count_Track  GenreId
0          1297        1
1           579        7
2           374        3
3           332        4
4           130        2
5            93       19
6            81        6
7            74       24
8            64       21
9            61       14
10           58        8
11           48        9
12           43       10
13           40       23
14           35       17
15           30       15
16           28       13
17           28       16
18           26       20
19           24       12
20           17       22
21           15       11
22           13       18
23           12        5
24            1       25

In [7]:
# Kiểm tra dữ liệu bảng "track"
query = """
        SELECT AlbumID, AVG(Milliseconds) as Mean_Time
        FROM track
        WHERE GenreID = 1
        
        GROUP BY AlbumID
        --LIMIT (5)
"""
cursor.execute(query)
fetch_data = cursor.fetchall()
pd.DataFrame(fetch_data, columns=[item[0] for item in cursor.description])

AlbumId      Mean_Time
0          1  240041.500000
1          2  342562.000000
2          3  286029.333333
3          4  306657.375000
4          5  294113.933333
..       ...            ...
112      246  262851.615385
113      252  310774.000000
114      256  351275.416667
115      257  287370.166667
116      265  205983.500000

[117 rows x 2 columns]

**Mỗi album có bao nhiêu track?**

In [9]:
query = """
        SELECT albumid, count(trackid) as count
        FROM track
        GROUP BY AlbumId        
        ORDER BY count DESC
        LIMIT (10)
"""
cursor.execute(query)
pd.DataFrame(cursor.fetchall(), columns=[item[0] for item in cursor.description])

AlbumId  count
0      141     57
1       23     34
2       73     30
3      229     26
4      230     25
5      251     25
6       83     24
7      231     24
8      253     24
9       24     23

**Thời gian trung bình (theo milisecond) của các track trong mỗi album**

In [13]:
query = """
        SELECT albumid, AVG(Milliseconds) as AVG_TIME
        FROM track
        GROUP BY AlbumId        
        ORDER BY AVG_TIME DESC
        LIMIT (10)
"""
cursor.execute(query)
pd.DataFrame(cursor.fetchall(), columns=[item[0] for item in cursor.description])

AlbumId      AVG_TIME
0      253  2.925574e+06
1      227  2.778265e+06
2      229  2.717907e+06
3      231  2.637068e+06
4      226  2.622250e+06
5      228  2.599142e+06
6      230  2.594197e+06
7      254  2.484567e+06
8      261  2.321673e+06
9      251  1.532684e+06

**Tổng dung lượng của các track trong từng album**

In [14]:
query = """
        SELECT albumid, SUM(Bytes) as sum_bytes, MIN(Bytes), MAX(bytes)
        FROM track
        GROUP BY AlbumId        
        ORDER BY sum_bytes DESC
        LIMIT (10)
"""
cursor.execute(query)
pd.DataFrame(cursor.fetchall(), columns=[item[0] for item in cursor.description])

AlbumId    sum_bytes  MIN(Bytes)  MAX(bytes)
0      229  13917603291   486675063  1059546140
1      253  12872621850   487899692   587051735
2      231  12344960921   457364940   574325829
3      228  11781321607   475996611   549353481
4      227  10059916535   462818231  1054423946
5      261   7708725642    20831818   526865050
6      251   7652731262   245378749   515301752
7      250   5711964665   244626927   327642458
8      230   5280909854   183867185   228896396
9      249   1610359572   257879716   290482361

**Group theo nhiều cột: mỗi loại media có bao nhiêu genre, trong đó có bao nhiêu track?**

In [15]:
query = """
        SELECT MediaTypeId, GenreId, count(trackid) as count
        FROM track
        WHERE MediaTypeId BETWEEN 3 and 5
        
        GROUP BY MediaTypeId, GenreId
        ORDER BY MediaTypeId ASC
        --LIMIT (10)
"""
cursor.execute(query)
pd.DataFrame(cursor.fetchall(), columns=[item[0] for item in cursor.description])

MediaTypeId  GenreId  count
0             3       18     13
1             3       19     93
2             3       20     26
3             3       21     64
4             3       22     17
5             3       23      1
6             4       23      1
7             4       24      6
8             5        1      2
9             5        2      3
10            5        7      1
11            5       15      2
12            5       16      2
13            5       24      1

## 1.2. Kết hợp `GROUP BY` và `HAVING`

**Chỉ lọc ra các track có GenreId là 2 hoặc MediaTypeId là 2**

In [23]:
query = """
        SELECT albumid, SUM(Bytes) as sum_bytes, MIN(Bytes), MAX(bytes)
        FROM track
        WHERE MediaTypeId=2
        
        GROUP BY AlbumId
        --HAVING sum_bytes>10000000000
        --HAVING GenreId=2
        HAVING sum_bytes>50000000
                
        ORDER BY sum_bytes DESC
        --LIMIT (10)
"""
cursor.execute(query)
pd.DataFrame(cursor.fetchall(), columns=[item[0] for item in cursor.description])

AlbumId  sum_bytes  MIN(Bytes)  MAX(bytes)
0      255   83470311     2447453     4656660
1       94   79242814     4383764     9367328
2       91   77516317     2550030    10201342
3      256   67857091     3071042     8858616
4      257   57769571     3482099     6491444
5      269   56198102     3892066    11157785
6      270   54019835     3355715     4606408
7       90   53625668     3301971     6687123
8      271   52304949     3484335     4901540

**Chỉ lọc ra các album có tổng thời gian nằm trong khoảng 500k - 1000k milisec**

In [24]:
query = """
        SELECT albumid, AVG(Milliseconds) as AVG_TIME, SUM(Milliseconds) as SUM_TIME
        FROM track
        
        GROUP BY AlbumId
        HAVING SUM_TIME BETWEEN 500000 and 1000000
        
        ORDER BY AVG_TIME DESC
        LIMIT (10)
"""
cursor.execute(query)
pd.DataFrame(cursor.fetchall(), columns=[item[0] for item in cursor.description])

AlbumId  AVG_TIME  SUM_TIME
0      294  596519.0    596519
1      279  582029.0    582029
2      330  567494.0    567494
3      312  561967.0    561967
4      301  560342.0    560342
5      299  545203.0    545203
6      311  526696.0    526696
7      292  522099.0    522099
8      273  501503.0    501503
9      173  347972.0    695944

# 2. Truy vấn dữ liệu từ nhiều tables

## 2.1. Các mệnh đề JOIN

In [27]:
query = """
        SELECT Name, album.Title
        FROM track
        JOIN album
            ON track.AlbumId = album.AlbumId
        --LIMIT (5)
"""
cursor.execute(query)
pd.DataFrame(cursor.fetchall(), columns=[item[0] for item in cursor.description])

Name  \
0               For Those About To Rock (We Salute You)   
1                                     Balls to the Wall   
2                                       Fast As a Shark   
3                                     Restless and Wild   
4                                  Princess of the Dawn   
...                                                 ...   
3498  Pini Di Roma (Pinien Von Rom) \ I Pini Della V...   
3499  String Quartet No. 12 in C Minor, D. 703 "Quar...   
3500               L'orfeo, Act 3, Sinfonia (Orchestra)   
3501  Quintet for Horn, Violin, 2 Violas, and Cello ...   
3502                                      Koyaanisqatsi   

                                                  Title  
0                 For Those About To Rock We Salute You  
1                                     Balls to the Wall  
2                                     Restless and Wild  
3                                     Restless and Wild  
4                                     Restless and Wild  
...                                                 ...  
3498                             Respighi:Pines of Rome  
3499  Schubert: The Late String Quartets & String Qu...  
3500                                Monteverdi: L'Orfeo  
3501                              Mozart: Chamber Music  
3502  Koyaanisqatsi (Soundtrack from the Motion Pict...  

[3503 rows x 2 columns]

**(1) Mỗi album là của artist nào?**

In [29]:
query = """
        SELECT title, name
        FROM album
        JOIN artist
            ON album.artistid = artist.artistid
        --LIMIT(10)
"""
cursor.execute(query)
pd.DataFrame(cursor.fetchall(), columns=[item[0] for item in cursor.description])

Title  \
0                For Those About To Rock We Salute You   
1                                    Balls to the Wall   
2                                    Restless and Wild   
3                                    Let There Be Rock   
4                                             Big Ones   
..                                                 ...   
342                             Respighi:Pines of Rome   
343  Schubert: The Late String Quartets & String Qu...   
344                                Monteverdi: L'Orfeo   
345                              Mozart: Chamber Music   
346  Koyaanisqatsi (Soundtrack from the Motion Pict...   

                                                  Name  
0                                                AC/DC  
1                                               Accept  
2                                               Accept  
3                                                AC/DC  
4                                            Aerosmith  
..                                                 ...  
342                                     Eugene Ormandy  
343                             Emerson String Quartet  
344  C. Monteverdi, Nigel Rogers - Chiaroscuro; Lon...  
345                                      Nash Ensemble  
346                              Philip Glass Ensemble  

[347 rows x 2 columns]

In [52]:
# Đặt tên viết tắt cho bảng truy vấn
query = """
        SELECT title as 'Album Title', name as 'Artist Name'
        FROM album as a1
        JOIN artist as a2
            ON a1.artistID = a2.artistID
        --LIMIT(10)
"""
cursor.execute(query)
df = pd.DataFrame(cursor.fetchall(), columns=[item[0] for item in cursor.description])
df

Album Title  \
0                For Those About To Rock We Salute You   
1                                    Balls to the Wall   
2                                    Restless and Wild   
3                                    Let There Be Rock   
4                                             Big Ones   
..                                                 ...   
342                             Respighi:Pines of Rome   
343  Schubert: The Late String Quartets & String Qu...   
344                                Monteverdi: L'Orfeo   
345                              Mozart: Chamber Music   
346  Koyaanisqatsi (Soundtrack from the Motion Pict...   

                                           Artist Name  
0                                                AC/DC  
1                                               Accept  
2                                               Accept  
3                                                AC/DC  
4                                            Aerosmith  
..                                                 ...  
342                                     Eugene Ormandy  
343                             Emerson String Quartet  
344  C. Monteverdi, Nigel Rogers - Chiaroscuro; Lon...  
345                                      Nash Ensemble  
346                              Philip Glass Ensemble  

[347 rows x 2 columns]

**Kiểm tra lại số lượng records của 2 bảng Album và Artist**

In [31]:
query = """
        SELECT COUNT(*)
        FROM album
"""
cursor.execute(query)
cursor.fetchone()

(347,)

In [32]:
query = """
        SELECT COUNT(*)
        FROM artist
"""
cursor.execute(query)
cursor.fetchone()

(275,)

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 347 entries, 0 to 346
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Album Title  347 non-null    object
 1   Artist Name  347 non-null    object
dtypes: object(2)
memory usage: 5.5+ KB


**==> Có những artist có vài album**

In [34]:
df.groupby('Artist Name').count().size

204

**==> Có tổng cộng 275 artist, nhưng ở đây chỉ có 204 artist có album**

**==> Có 71 artist không có album nào**

**(2) Mỗi artist có những albums nào?**

In [51]:
# Thử với JOIN, INNER JOIN, CROSS JOIN, LEFT JOIN
query = """
        SELECT ar.name AS 'Artist Name', al.title AS 'Album Title'
        FROM Artist ar
        LEFT JOIN Album al 
            USING(artistid)
"""
cursor.execute(query)
df = pd.DataFrame(cursor.fetchall(), columns=[item[0] for item in cursor.description])
# df
df.sample(20)
# df.isnull().sum()
# df[df.isnull().any(axis=1)]

Artist Name  \
366          Leonard Bernstein & New York Philharmonic   
342  Richard Marlow & The Choir of Trinity College,...   
287                                         Whitesnake   
320                                  Charlie Brown Jr.   
175                                        Iron Maiden   
42                                        Led Zeppelin   
305                                       Jack Johnson   
6                                      Alice In Chains   
160                                        Iron Maiden   
337                                     Aaron Goldberg   
309                                 The Postal Service   
106                                        Deep Purple   
273                                                 U2   
293                     Battlestar Galactica (Classic)   
94                                          Spyro Gyra   
295                Christina Aguilera featuring BigElf   
350  Academy of St. Martin in the Fields & Sir Nevi...   
198                                          Motörhead   
63                                             O Rappa   
326                                        Nação Zumbi   

                                           Album Title  
366  Great Performances - Barber's Adagio and Other...  
342                                  Allegri: Miserere  
287                                               None  
320                                               None  
175                                  Somewhere in Time  
42                                     Led Zeppelin II  
305                                               None  
6                                             Facelift  
160                                    A Real Live One  
337                                             Worlds  
309                                               None  
106                                       Machine Head  
273                                  B-Sides 1980-1990  
293           Battlestar Galactica (Classic), Season 1  
94                                       Morning Dance  
295                                               None  
350                  The World of Classical Favourites  
198                                      Ace Of Spades  
63   Radio Brasil (O Som da Jovem Vanguarda) - Sele...  
326                                               None

**(3) 10 artists có số album nhiều nhất / ít nhất?**

In [66]:
df.groupby('Artist Name').count().nlargest(columns='Album Title', n = 10)

Album Title
Artist Name               
Iron Maiden             21
Led Zeppelin            14
Deep Purple             11
Metallica               10
U2                      10
Ozzy Osbourne            6
Pearl Jam                5
Faith No More            4
Foo Fighters             4
Lost                     4

In [69]:
df.groupby('Artist Name').count().sort_values('Album Title')

Album Title
Artist Name                                                    
A Cor Do Som                                                  0
Pedro Luís & A Parede                                         0
Pedro Luís E A Parede                                         0
Peter Tosh                                                    0
Edson, DJ Marky & DJ Patife Featuring Fernanda ...            0
...                                                         ...
Metallica                                                    10
U2                                                           10
Deep Purple                                                  11
Led Zeppelin                                                 14
Iron Maiden                                                  21

[275 rows x 1 columns]

## 2.2. Kết hợp các lệnh khác trong SQL

**(1) Kết hợp ORDER BY**

In [70]:
query = """
        SELECT ar.name AS 'Artist Name', al.title AS 'Album Title'
        FROM Artist ar
        LEFT JOIN Album al
            ON ar.artistid = al.artistid
        ORDER BY ar.name ASC
"""
cursor.execute(query)
df = pd.DataFrame(cursor.fetchall(), columns=[item[0] for item in cursor.description])
df

Artist Name  \
0                                 A Cor Do Som   
1                                        AC/DC   
2                                        AC/DC   
3    Aaron Copland & London Symphony Orchestra   
4                               Aaron Goldberg   
..                                         ...   
413                                        Xis   
414                             Yehudi Menuhin   
415                                   Yo-Yo Ma   
416                             Youssou N'Dour   
417                             Zeca Pagodinho   

                               Album Title  
0                                     None  
1    For Those About To Rock We Salute You  
2                        Let There Be Rock  
3            A Copland Celebration, Vol. I  
4                                   Worlds  
..                                     ...  
413                                   None  
414       Bartok: Violin & Viola Concertos  
415                 Bach: The Cello Suites  
416                                   None  
417                       Ao Vivo [IMPORT]  

[418 rows x 2 columns]

**(2) Kết hợp GROUP BY**

**Mỗi artist có bao nhiêu album? 10 artist có số album nhiều nhất?**

In [88]:
# Cách 1: Dùng lệnh count trực tiếp
query = """
        SELECT ar.name AS 'Artist Name', count(al.artistid) as NoAlbums
        FROM Artist ar
        LEFT JOIN Album al
            ON ar.artistid = al.artistid
        
        GROUP BY ar.name
        --HAVING NoAlbums > 5
        ORDER BY NoAlbums DESC
        LIMIT (10)
"""
cursor.execute(query)
df = pd.DataFrame(cursor.fetchall(), columns=[item[0] for item in cursor.description])
df

Artist Name  NoAlbums
0      Iron Maiden        21
1     Led Zeppelin        14
2      Deep Purple        11
3               U2        10
4        Metallica        10
5    Ozzy Osbourne         6
6        Pearl Jam         5
7  Various Artists         4
8        Van Halen         4
9             Lost         4

In [89]:
# Cách 2: Lồng ghép kết quả SELECT vào 1 SELECT khác
query = """
        SELECT ar.name, ar1.count
        FROM
            (SELECT al.artistid, count(al.albumid) as count
            FROM Album al
            GROUP BY al.artistid
            ORDER BY count DESC) ar1
        JOIN artist ar
            ON ar.artistid = ar1.artistid
        LIMIT(10)
"""
cursor.execute(query)
df = pd.DataFrame(cursor.fetchall(), columns=[item[0] for item in cursor.description])
df

Name  count
0      Iron Maiden     21
1     Led Zeppelin     14
2      Deep Purple     11
3        Metallica     10
4               U2     10
5    Ozzy Osbourne      6
6        Pearl Jam      5
7  Various Artists      4
8    Faith No More      4
9     Foo Fighters      4

In [92]:
query = """ 
        SELECT a1.Name, count
        FROM
            (SELECT al.artistid, count(al.albumid) as count
            FROM Album al
            GROUP BY al.artistid
            ORDER BY count DESC
            LIMIT (10)
            ) t1
        
        JOIN artist as a1
            ON t1.artistid = a1.artistid
        
        WHERE count > 10
"""
cursor.execute(query)
df = pd.DataFrame(cursor.fetchall(), columns=[item[0] for item in cursor.description])
df

Name  count
0   Iron Maiden     21
1  Led Zeppelin     14
2   Deep Purple     11

**Tên của 10 album có số track nhiều nhất?**

In [94]:
query = """
        SELECT album.title, count
        FROM
            (SELECT albumid, count(trackid) as count
            FROM track
            GROUP BY AlbumId        
            ORDER BY count DESC
            LIMIT (10)) AS tr
        JOIN album
            ON album.albumid = tr.albumid
"""
cursor.execute(query)
pd.DataFrame(cursor.fetchall(), columns=[item[0] for item in cursor.description])

Title  count
0                               Greatest Hits     57
1                              Minha Historia     34
2                                   Unplugged     30
3                              Lost, Season 3     26
4                              Lost, Season 1     25
5                        The Office, Season 3     25
6  My Way: The Best Of Frank Sinatra [Disc 1]     24
7                              Lost, Season 2     24
8    Battlestar Galactica (Classic), Season 1     24
9                              Afrociberdelia     23

## 2.3. Các mệnh đề UNION

In [95]:
query = """
        SELECT LastName, FirstName, 'Emp' AS Type
        FROM employee
        --LIMIT (5)
        
        UNION
        SELECT LastName, FirstName, 'Cus'
        FROM Customer
        ORDER BY Type DESC
        
        --LIMIT (5)
"""
cursor.execute(query)
pd.DataFrame(cursor.fetchall(), columns=[item[0] for item in cursor.description])

LastName  FirstName Type
0          Adams     Andrew  Emp
1       Callahan      Laura  Emp
2        Edwards      Nancy  Emp
3        Johnson      Steve  Emp
4           King     Robert  Emp
..           ...        ...  ...
62      Tremblay   François  Cus
63  Van der Berg   Johannes  Cus
64   Wichterlová  František  Cus
65        Wójcik  Stanisław  Cus
66    Zimmermann       Fynn  Cus

[67 rows x 3 columns]

# 3. Demo: NoSQL

**Tạo một database sử dụng kiểu `dict`**

In [3]:
{'danh_sach': [{'ten': "Nam", 'tuoi': 23}, {'ten': "Lan", 'tuoi': 26}]}

{'danh_sach': [{'ten': 'Nam', 'tuoi': 23}, {'ten': 'Lan', 'tuoi': 26}]}

In [9]:
db = {'lop': {'id': 7, 'ten': 'Data Science',
              'giang-vien': {'id':5, 'ten': 'De-Thu', 'tuoi': 33},
              'danh-sach':[{'hoc-vien': {'id':3, 'ten': 'Nam'}},
                           {'hoc-vien': {'id':5, 'ten': 'Lan'}}
                          ]
             }
     }
db

{'lop': {'id': 7,
  'ten': 'Data Science',
  'giang-vien': {'id': 5, 'ten': 'De-Thu', 'tuoi': 33},
  'danh-sach': [{'hoc-vien': {'id': 3, 'ten': 'Nam'}},
   {'hoc-vien': {'id': 5, 'ten': 'Lan'}}]}}

**Bản thân CSDL này là một dict với cặp key-value**

In [6]:
db.keys()

dict_keys(['lop'])

In [10]:
db.values()

dict_values([{'id': 7, 'ten': 'Data Science', 'giang-vien': {'id': 5, 'ten': 'De-Thu', 'tuoi': 33}, 'danh-sach': [{'hoc-vien': {'id': 3, 'ten': 'Nam'}}, {'hoc-vien': {'id': 5, 'ten': 'Lan'}}]}])

In [11]:
db['lop']['giang-vien']['ten']

'De-Thu'

In [13]:
db['lop']['danh-sach'][0]

{'hoc-vien': {'id': 3, 'ten': 'Nam'}}

In [14]:
print(db['lop'].get('giang-vien'))

{'id': 5, 'ten': 'De-Thu', 'tuoi': 33}


In [15]:
print(db['lop'].get('mentor'))

None
